packages/iov-ethereum/src/ethereumconnection.ts

```diff
@@ -196,7 +233,62 @@ export class EthereumConnection implements BcpConnection {
   }
 
   public watchBlockHeaders(): Stream<BlockHeader> {
-    throw new Error("Not implemented");
+    // see https://github.com/ethereum/go-ethereum/wiki/RPC-PUB-SUB#newheads
+    this.socket!.send(
+      JSON.stringify({
+        id: "watchBlockHeaders",
+        jsonrpc: "2.0",
+        method: "eth_subscribe",
+        params: ["newHeads", { includeTransactions: true }],
+      }),
+    );
+
+    let headersSubscription: Subscription;
+    const producer: Producer<BlockHeader> = {
+      start: listener => {
+        headersSubscription = this.socket!.events.subscribe({
+          next: header => {
+            const blockHeaderJson = JSON.parse(header.data);
+            if (blockHeaderJson.method === "eth_subscription") {
+              const listening = getListenerBySubscription(blockHeaderJson.params.subscription);
+              if (listening) {
+                // Missed transaction count in newHeads subscription
+                this.getBlockHeader(decodeHexQuantity(blockHeaderJson.params.result.number))
+                  .then(blockHeader => listener.next(blockHeader))
+                  .catch(error => listener.error(error));
+              }
+            } else if (blockHeaderJson.id) {
+              // This is the response from eth server when subscribing to an event
+              // store eth subscription id to unsubscribe later
+              if (typeof blockHeaderJson.result === "string") {
+                addToListeners(blockHeaderJson.id, blockHeaderJson.result);
+              } else if (blockHeaderJson.result === true) {
+                removeFromListeners(blockHeaderJson.id);
+              }
+            }
+          },
+          error: error => listener.error(error),
+          complete: () => listener.error("Source stream completed"),
+        });
+      },
+      stop: () => {
+        headersSubscription.unsubscribe();
+      },
+    };
+    return Stream.create(producer);
+  }
+
+  public unwatchBlockHeaders(): void {
```

Comment:
All the logic from here belongs to the `stop` block of the producer. That way you don't need to call `unwatchBlockHeaders()` explicitly but just unsubscribe from the stream.

Reply:
Sorry, but I need to see the full updated code and error to debug. This streaming logic is not easy to follow.

Reply:
if you console.log `public disconnect()` and `public watchBlockHeaders() stop`, you will see in the output that `disconnect()` executes before `watchBlockHeaders() stop`, even if they are called in the right order: first stop (unsubscribe) and then disconnect.
So any logic inside `stop`, related with sockets, wont work.
Do you want me to push the error into CI? (it will be "Failed: Websocket is not open")

Reply:
Yeah I guess the issue is in the test logic (see different comment)

Reply:
got it!